ipa-core/src/protocol/dp/mod.rs

```diff
@@ -268,9 +275,149 @@ where
 
             Ok(noisy_histogram)
         }
+        DpMechanism::DiscreteLaplace { epsilon } => {
+            let noise_params = NoiseParams {
+                epsilon,
+                per_user_credit_cap: 2_u32.pow(u32::try_from(SS_BITS).unwrap()),
+                ..Default::default()
+            };
+
+            let truncated_discret_laplace = OPRFPaddingDp::new(
+                noise_params.epsilon,
+                noise_params.delta,
+                noise_params.per_user_credit_cap,
+            )?;
+
+            assert!((epsilon - noise_params.epsilon).abs() < 0.001);
+            let (mean, _) = truncated_discret_laplace.mean_and_std();
+            tracing::info!(
+                "In dp_for_histogram with Truncated Discrete Laplace noise: \
+                epsilon = {epsilon}, \
+                delta = {}, \
+                per_user_credit_cap = {}, \
+                noise mean (including all three pairs of noise) = {}, \
+                OV::BITS = {}",
+                noise_params.delta,
+                noise_params.per_user_credit_cap,
+                mean * 3.0,
+                OV::BITS,
+            );
+
+            let noised_output = apply_laplace_noise_pass::<C, OV, B>(
+                &ctx.narrow(&DPStep::LaplacePass1),
+                histogram_bin_values,
+                Role::H1,
+                &noise_params,
+            )
+            .await?;
+
+            let noised_output = apply_laplace_noise_pass::<C, OV, B>(
+                &ctx.narrow(&DPStep::LaplacePass2),
+                noised_output,
+                Role::H2,
+                &noise_params,
+            )
+            .await?;
+
+            let noised_output = apply_laplace_noise_pass::<C, OV, B>(
+                &ctx.narrow(&DPStep::LaplacePass3),
+                noised_output,
+                Role::H3,
+                &noise_params,
+            )
+            .await?;
+
+            Ok(Vec::transposed_from(&noised_output)?)
+        }
     }
 }
 
+/// # Errors
+/// will propagate errors from constructing a `truncated_discrete_laplace` distribution.
+/// # Panics
+///
+pub async fn apply_laplace_noise_pass<C, OV, const B: usize>(
+    ctx: &C,
+    histogram_bin_values: BitDecomposed<Replicated<Boolean, B>>,
+    excluded_helper: Role,
+    noise_params: &NoiseParams,
+) -> Result<BitDecomposed<Replicated<Boolean, B>>, Error>
+where
+    C: Context,
+    OV: BooleanArray + U128Conversions,
+    Boolean: Vectorizable<B> + FieldSimd<B>,
+    Replicated<Boolean, B>: BooleanProtocols<C, B>,
+    BitDecomposed<AdditiveShare<Boolean, B>>:
+        for<'a> TransposeFrom<&'a [AdditiveShare<OV>; B], Error = Infallible>,
+    AdditiveShare<OV>: ReplicatedSecretSharing<OV>,
+{
+    let mut noise_values = vec![];
```

Comment:
probably cleaner to use `Vec::with_capacity(B);`, since we know the size. @akoshelev can you double check me here?
```suggestion
    let mut noise_values = Vec::with_capacity(B);
```

Reply:
array approach makes sense to me, given that we ultimately need an array anyway

Reply:
we probably should split this into smaller functions as well

Reply:
merged Erik's refactor to build array directly